core/src/Streamly/Internal/Data/Serialize.hs

```diff
@@ -0,0 +1,99 @@
+module Streamly.Internal.Data.Serialize
+    ( Size(..)
+    , Serialize(..)
+    ) where
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Control.Monad (void)
+import Data.List (foldl')
+import Data.Proxy (Proxy (..))
+import Streamly.Internal.Data.Unboxed (MutableByteArray(..))
+
+import GHC.Int (Int16(..), Int32(..), Int64(..), Int8(..))
+import GHC.Word (Word16(..), Word32(..), Word64(..), Word8(..))
+import GHC.Stable (StablePtr(..))
+
+import qualified Streamly.Internal.Data.Unboxed as Unbox
+
+import GHC.Exts
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data Size a
+    = VarSize (a -> Int)
+    | ConstSize (Proxy a -> Int)
+
+class Serialize a where
+    size :: Size a
+    peekByteIndex :: Int -> MutableByteArray -> IO (Int, a)
```

Comment:
Need to document the APIs.

Also, we can put a dev comment here. We can implement this without returning the `Int` offset but that may require traversing the Haskell structure again to get the size. Therefore, this is a performance optimization.

Reply:
We can name these `deserialize` and `serialize` to avoid naming conflict with Unbox?